[Navigation screen] Replace component state and refs with Redux state

adamziel opened the pull request:
## Description

This PR replaces the component state like `const [blocks, setBlocks] = useState();` and `const menuItemRef = useRef()` in favor of the redux state and data layer.

## How has this been tested?
1. Enable the experimental navigation screen
1. Go to `/wp-admin/admin.php?page=gutenberg-navigation` and interact with it in every way you can think of: create a new menu, add menu items, edit existing ones, remove some, save the menu, rinse and repeat without refreshing the page.
1. Confirm everything worked as expected.

## Types of changes
New feature

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

adamziel commented on packages/edit-navigation/src/components/menu-editor/index.js:
```diff
@@ -6,66 +6,81 @@ import {
-		menuItems
-	);
-	const saveMenuItems = () => eventuallySaveMenuItems( blocks, menuItemsRef );
+	const postId = useStubPost( query );
```
Instead of keeping blocks in state, this PR experiments with creating a "stub" post that serves as container for all our blocks but is never meant to be saved.

adamziel commented on packages/edit-navigation/src/components/menu-editor/index.js:
```diff
@@ -6,66 +6,81 @@ import {
+	onDeleteMenu,
+} ) => {
+	const isLargeViewport = useViewportMatch( 'medium' );
+	const [
+		blocks,
+		onInput,
+		onChange,
+		saveMenuItems,
+	] = useNavigationBlockEditor( query, postId );
```
Instead of having custom `block`/`onInput`/`onChange` handlers, this uses the same ones as the regular post editor.

adamziel commented on packages/edit-navigation/src/components/menu-editor/promise-queue.js:
```diff
@@ -7,15 +7,28 @@ export default class PromiseQueue {
+		this.halted = false;
+	}
+
+	halt() {
+		this.halted = true;
+		this.queue = [];
+		this.listeners = [];
```
Remember how new menu items are created whenever a new block is added? Changing an active menu should stop any further processing of these if some are in progress. My idea in this PR is to simply "halt" a promise queue which means no further actions will be processed an no listeners will be notified - eventually the entire thing will be just garbage collected.

adamziel commented on packages/edit-navigation/src/components/menu-editor/use-create-missing-menu-items.js:
```diff
@@ -17,41 +18,95 @@ import PromiseQueue from './promise-queue';
+	}
+}
+
+function useStoreSavedMenuItem( query ) {
+	const { assignMenuItemIdToClientId } = useDispatch(
+		'core/edit-navigation'
+	);
+	const { receiveEntityRecords } = useDispatch( 'core' );
+	const select = useSelect( ( s ) => s );
+	return useCallback(
+		( clientId, menuItem ) => {
+			assignMenuItemIdToClientId( query, menuItem.id, clientId );
+			receiveEntityRecords(
+				'root',
+				'menuItem',
+				[ ...select( 'core' ).getMenuItems( query ), menuItem ],
+				query,
+				false
+			);
+		},
+		[ query ]
+	);
```
I am not too happy about this part. Previously `menuItems` were kept in react ref. Now the core data layer is the primary source of truth. This means that creating a temporary menu item should be reflected in redux state. The only way to do it ATM is `receiveEntityRecords` which accepts an entire page of results. Maybe we should add an action like `appendEntityRecord` to avoid this `select` madness here?

adamziel commented on packages/edit-navigation/src/components/menu-editor/use-navigation-block-editor.js:
```diff
@@ -0,0 +1,83 @@
+	);
+	const select = useSelect( ( s ) => s );
+	const saveBlocks = async ( blocks ) => {
+		const menuItemsByClientId = mapMenuItemsByClientId(
+			select( 'core' ).getMenuItems( query ),
+			select( 'core/edit-navigation' ).getMenuItemIdToClientIdMapping(
+				query
+			)
+		);
```
`saveBlocks` is not called synchronously when the save button is pressed. Instead, we wait for all draft menu items to be created first, and only then `saveBlocks` is called. This means that we need a fresh version of `menuItemId -> clientId` mapping - if we simply selected it in line 42 above then it would be outdated by the time this function is called. Ditto for `getMenuItems`. I am not too happy about this solution, but the only alternative I can think of that would involve a "valid" use of `useSelect` calls is keeping a state variable like `const [shouldSave, setShouldSave] = useState( false );`. Then, when all the async operations are finished, we'd re-render the component using `setShouldSave( true );`, and react to this change in an effect which would then have access to the most recent version of `select()`-ed data.

adamziel commented on packages/edit-navigation/src/components/menu-editor/use-stub-post.js:
```diff
@@ -0,0 +1,77 @@
+		);
+		setMenuItemsToClientIdMapping( query, menuItemIdToClientId );
+
+		const post = createStubPost( query.menus, navigationBlock );
+		const entityStored = receiveEntityRecords(
+			'root',
+			'postType',
+			post,
+			{ id: post.id },
+			false
+		);
+		entityStored.then( () => {
+			setPostId( post.id );
+		} );
```
I am pretty sure this isn't the correct way to store this stub post. Should this even be a post? Maybe a dedicated entity kind would work better?

adamziel commented on packages/edit-navigation/src/store/index.js:
```diff
@@ -0,0 +1,36 @@
+ */
+const MODULE_KEY = 'core/edit-navigation';
+
+/**
+ * Block editor data store configuration.
+ *
+ * @see https://github.com/WordPress/gutenberg/blob/master/packages/data/README.md#registerStore
+ *
+ * @type {Object}
+ */
```
I didn't pay too much attention to the comments yet, let it be - I will clean it all up later :-)

draganescu commented:
cc @noisysocks

noisysocks commented:
I am thinking we need to implement something that demonstrates the value of this refactor in order to gauge whether the extra bit of complexity is worth it or not. Could you maybe add a rough pass at undo/redo to this PR to see how it looks? That's what I imagine the primary value of using `@wordpress/data` is.

talldan commented:
Fixing https://github.com/WordPress/gutenberg/issues/22625 could be a good first implementation of an `edit-navigation` store as well—moving the selected menu id from state to store would solve that.

noisysocks commented:
The functions in `wp.data.select( '...' )` and `wp.data.dispatch( '...' )` are callable by third parties and I generally like to think of them as official APIs for all WordPress developers to use.

With that in mind, it's weird to me that this PR has `wp.data.select( 'core/edit-navigation' ).getMenuItemIdToClientIdMapping()`, `wp.data.dispatch( 'core/edit-navigation' ).setMenuItemsToClientIdMapping()`, and `wp.data.dispatch( 'core/edit-navigation' ).assignMenuItemIdToClientId()`. These functions are awfully specific. They expose the concept of _mapping menu item IDs to block client IDs_ which is an implementation detail that I don't think will be useful outside of our needs.

Could we move the code that's in `useNavigationBlockEditor` into selectors and actions that define a concrete API for working with block-based menus? How these selectors and actions convert blocks to and from menu items would then be an implementation detail. I don't think (but am not 100%!) that there's anything in `useNavigationBlockEditor` that can't be done using actions which use `select()` and `apiFetch()` controls.

To illustrate, here's what I imagine you'd be able to do in the DevTools console:

```js
// Hits GET /wp/v2/menu-items, munges the data into Link blocks, and returns
// everything as a fake post.
const post = wp.data.select( 'edit-navigation' ).getNavigationPost( /* menuId: */ 123 ); 
console.log( post.id ); // navigation-post-123
console.log( post.blocks ); // [ { name: 'core/navigation', ... } ]

// Hits POST /wp/v2/menu-items once for every Link block that doesn't have an
// associated menu item. (IDK what a good name for this is.)
wp.data.dispatch( 'edit-navigation' ).createMissingMenuItems( post );

// Hits the Customiser endpoint and saves everything.
wp.data.dispatch( 'edit-navigation' ).saveNavigationPost( post );

// If we put our fake post into the 'core' registry then we can do things like
// undo and redo.
wp.data.dispatch( 'core' ).receiveEntityRecords( 'postType', 'post', [ post ] );
wp.data.dispatch( 'core' ).editEntityRecord(
	'postType',
	'post',
	'navigation-post-123',
	{ blocks: [
		...post.blocks,
		{ name: 'core/paragraph', content: 'This will be undone!', ... }
	] }
);
wp.data.dispatch( 'core' ).undo();
wp.data.dispatch( 'core' ).redo();
```

Does that make sense? Thoughts?

adamziel commented:
> The functions in wp.data.select( '...' ) and wp.data.dispatch( '...' ) are callable by third parties and I generally like to think of them as official APIs for all WordPress developers to use.

TIL - that's a good mental model, thank you!

> They expose the concept of mapping menu item IDs to block client IDs which is an implementation detail that I don't think will be useful outside of our needs.

@noisysocks while I agree the name `getMenuItemIdToClientIdMapping()` is awful, we need to store that mapping somewhere. At the moment we use a ref which works fine but doesn't really allow us to expose an API like `wp.data.select( 'core/edit-navigation' ).getMenuItem( clientId )` which is what you mentioned in the following comment: https://github.com/WordPress/gutenberg/issues/22022#issuecomment-631143375

My understanding is that in order to expose a `getMenuItem()` selector, the `clientId<->menuId` mapping has to be stored in redux state. This in turn means that we need actions and selectors to manage it.

An alternative approach would be storing `menuItemId` in block attributes - this should remove the need for any mapping-related actions and selectors while still making it possible to expose a `getMenuItem( clientId )` selector - what do you think?

talldan commented:
@adamziel I think you're on the right path storing this in state, but  a lot the comments were probably about making the public facing API friendlier rather than using an alternate solution like block attributes.

The block editor store has a lot of relevance in that there are reducers that encapsulate things like mappings (block order, parent/child relationships), but they're not necessarily exposed to the user directly through actions. For example `insertBlock` will update the `order`,  and `parents`. Similarly a lot of the commonly used selectors are a bit more higher-level than the reducers they read from, though with selectors there does tend to be a lot more composition.

It might be that we have an `updateMenuItem` action. Separately a reducer for storing the mapping between blocks and menu items might have a case for that action. `getMenuItemIdForBlock` or `getClientIdForMenuItem` also sound like reasonable selectors.

In terms of `saveNavigationPost`, not sure if we'd want to go for something like a generator action like the `editor` store has for `savePost`:
https://github.com/WordPress/gutenberg/blob/master/packages/editor/src/store/actions.js#L231

That's another way to encapsulate the various working parts, but lets have more conversation on that.

draganescu commented:
> Fixing #22625 could be a good first implementation of an edit-navigation store as well—moving the selected menu id from state to store would solve that.

@talldan @adamziel #22428 fixes that, because when we delete an entity we don't need to use state anymore. I think @noisysocks 's suggestion with demo-ing undo/redo in the navigation screen is a better demonstration of this PR's powers!

adamziel commented:
@noisysocks I explored getting rid of most hooks in favor of the actions, reducers, and such. Interestingly enough it ended up being quite similar to the API you proposed. I played with undo/redo and it seems to work 🎉 

Even better - there is no more weirdness with passing `select` to nested callbacks in react hooks.

We also now have a selector called `getMenuItemForClientId()` which does exactly what the name says. Implementing an inverse selector should be pretty easy too.

I also ended up moving the mapping to `post.meta` - if we treat the post as main model for the navigation then it makes sense to keep that data there. 

There are a few minor bugs that I noticed, but let's have more discussion about the approach before I go ahead and start fixing them.

adamziel commented on packages/edit-navigation/src/store/actions.js:
```diff
@@ -0,0 +1,251 @@
+	}
+}
+
+function serializeProcessing( callback ) {
```
This wrapper guarantees serial execution of data processing actions. `saveNavigationPost()` needs to wait for all the missing items to be created, and running multiple `createMissingMenuItems()` concurrently could result in sending more requests than required.

adamziel commented on packages/edit-navigation/src/store/actions.js:
```diff
@@ -0,0 +1,251 @@
+			}
+		);
+	} catch ( e ) {
+		yield dispatch(
+			'core/notices',
+			'createErrorNotice',
+			__( 'There was an error.' ),
+			{
+				type: 'snackbar',
+			}
+		);
```
It would be nice to extract these notices outside of the save action, but that would require some way to wait for `serialProcessing()` to finish. I think it could wait until another iteration.

adamziel commented on packages/edit-navigation/src/store/actions.js:
```diff
@@ -0,0 +1,251 @@
+	);
+
+	try {
+		yield* batchSave( menuId, menuItemsByClientId, post.blocks[ 0 ] );
```
This try catch needs more attention to make sure all the errors are handled.

talldan commented:
@draganescu You might be thinking of a different issue. https://github.com/WordPress/gutenberg/issues/22625 is related to the `menuId`, `setMenuId`. It doesn't look like #22428 touches that code.

noisysocks commented on packages/edit-navigation/src/store/actions.js:
```diff
@@ -0,0 +1,251 @@
+			return { status: 'pending' };
+		}
+
+		yield dispatch(
+			'core/edit-navigation',
+			'startProcessingMenuItems',
+			menuId
+		);
```
These actions don't strike me as particularly useful for third parties. We can emulate "private actions" by dispatching the action object directly here.

```js
return {
	type: 'START_PROCESSING_MENU_ITEMS',
	menuId,
};
```

adamziel replied:
Good idea!

noisysocks commented on packages/edit-navigation/src/store/actions.js:
```diff
@@ -0,0 +1,251 @@
+				'core/edit-navigation',
+				'enqueueAfterProcessing',
+				menuId,
+				callback
```
Should we be storing non-serialisable things like functions in a Redux store? I imagine it may break tooling such as the Redux DevTools.

adamziel replied:
We already store non-serializable things in other stores, e.g. `EquivalentKeyMap` instances. It's not a perfect solution, but it's good enough for starters. If any serialization issue comes up, it's entirely okay to just skip the `processing` part of the state tree.

noisysocks commented on packages/edit-navigation/src/store/index.js:
```diff
@@ -0,0 +1,40 @@
+
+const store = registerStore( MODULE_KEY, {
+	...storeConfig,
+	persist: [ 'preferences' ],
```
There is no `preferences` key in the reducer.

noisysocks commented on packages/edit-navigation/src/store/utils.js:
```diff
@@ -0,0 +1,20 @@
+export const buildNavigationPostId = ( menuId ) =>
+	`navigation-post-${ menuId }`;
+
+export function uuidv4() {
```
Should we use the `uuid` package instead? We already are using it in some packages e.g. `@wordpress/blocks`.

noisysocks commented:
> @talldan I also ended up moving the mapping to `post.meta` - if we treat the post as main model for the editor then it makes sense to keep that data there.

That's interesting! What are the pros and cons for doing this versus storing the mapping in the `edit-navigation` store?

noisysocks reviewed (commented):
Heck yeah! Nice work. I'm feeling pretty good about this. It's exactly what I'd pictured, only with very clever Adam-isms like `serializeProcessing()` 😄 

A few more notes in addition to the ones I wrote inline:

- This PR makes the terminology we use in `edit-navigation` a little all over the place. For example, we call the entity a "menu" in some function names (`MenuEditor`) but then a "navigation" in others (`useNavigationBlockEditor()`). We need to settle on a few key terms. I think what probably makes sense is: the _navigation screen_ edits a _navigation post_, and the _navigation store_ saves a _navigation post_ as a collection of _menus_ and _menu items_.

  - Semi-relatedly: is there anything we can do to make it clearer that a _navigation post_ is not a "real" post that can be persisted and accessed via `/wp/v2/posts`?

- I think, as much as possible, let's make sure that consumers of `wp.data.select( 'core/edit-navigation' )` and `wp.data.dispatch( 'core/edit-navigation')` don't need to care about _menu items_. They're an internal detail. For example, let's not export actions like `startProcessingMenuItems()`.

- Since it's now easier to do, should we add undo/redo to the Navigation screen? Even just having the keyboard shortcuts would be useful.

Let me know when you're ready for me to review this with a finer comb.

adamziel committed:
Squash

adamziel committed:
menu editor -> navigation editor

adamziel committed:
Rename CSS classes and shortcuts components to reflect menu vs navigation

adamziel committed:
"Privatize" concurrency-related actions

adamziel committed:
Migrate to uuid module

adamziel committed:
Remove persist: [ 'preferences' ]

adamziel commented on packages/edit-navigation/src/store/actions.js:
```diff
@@ -0,0 +1,251 @@
+				'receiveEntityRecords',
+				'root',
+				'menuItem',
+				[ ...menuItems, menuItem ],
```
We want to:

1. Append another entry to the store
1. Append entry ID to stored query results

Using just `menuItem` does 1) but breaks 2) as instead of appending it replaces the first ID. Using `[ ...menuItems, menuItem ]` both appends the new entry to `queriedData.items` AND appends entry ID to stored query results to `queriedData.queries["menus=123"]`.

adamziel committed:
Remove processing-related selectors in favor of controls which are not a public API

adamziel committed:
Wrap pending action in serializeProcessing()

adamziel committed:
Rename processing to processingQueue

adamziel committed:
Move the mapping back to state

adamziel committed:
Remove general resolveSelect control

adamziel committed:
Add undo/redo shortcuts

adamziel commented:
I addressed all your feedback @noisysocks, this is ready for another round.

> Semi-relatedly: is there anything we can do to make it clearer that a navigation post is not a "real" post that can be persisted and accessed via /wp/v2/posts?

Probably not much other than naming. `UnsavableNavigationPost`? `StubNavigationPost`? `InMemoryNavigationPost`?

adamziel commented on packages/edit-navigation/src/store/controls.js:
```diff
@@ -0,0 +1,138 @@
+	};
+}
+
+export function getMenuItemToClientIdMapping( postId ) {
+	return {
+		type: 'GET_MENU_ITEM_TO_CLIENT_ID_MAPPING',
+		postId,
+	};
+}
```
I ended up moving this mapping back to state - keeping it in the post turned out to be troublesome (e.g. every update created additional undo actions). I'm not sure if there is any other way to have a "private selector" other than custom control - I'm open to refactoring this bit.

adamziel commented on packages/edit-navigation/src/store/controls.js:
```diff
@@ -0,0 +1,138 @@
+	};
+}
+
+export function getNavigationPost( menuId ) {
```
`getNavigationPost` seems like the argument is a post id. Would `getNavigationPostByMenuId` be clearer?

noisysocks replied:
It should match the name of the selector. I'm not against renaming the selector to something like `getNavigationPostForMenu( menuId )`, though, if you want! 🙂

adamziel committed:
Remove persist option in proper store

adamziel commented on packages/edit-navigation/src/store/selectors.js:
```diff
@@ -0,0 +1,36 @@
+export const isResolvingNavigationPost = ( state, menuId ) => {
+	const post = getNavigationPost( state, menuId );
+	return ! post || post.meta.isResolving;
+};
```
Storing `isResolving` in meta is not ideal, keeping a proper "resolution" state could potentially solve the rogue request for the fake post.

adamziel commented:
Also, I updated the naming to be consistent about menu item vs navigator post to a certain extent. I don't want to pollute this PR with too many naming changes so let's do another one for any other changes of this kind.

adamziel committed:
Use the canonical way of keeping track of post resolution (and fix the rogue http query)

adamziel committed:
Add comments

adamziel commented:
I solved the issue with rogue http requests and this is now fully functional and ready for review. I don't like how we have multiple spinners now, but let's solve it in another PR - this one is already large enough.

adamziel committed:
Enforce menuId to be string

adamziel committed:
Add unit tests for mapping reducer

adamziel committed:
Add unit tests for processingQueue reducer

adamziel commented:
I added a few tests for starters, more will come in this or another PR

noisysocks commented on packages/edit-navigation/src/store/actions.js:
```diff
@@ -0,0 +1,274 @@
+}
+
+function* batchSave( menuId, menuItemsByClientId, navigationBlock ) {
+	const { nonce, stylesheet } = yield apiFetch( {
```
Need to handle any error that this request may have.

adamziel replied:
There are no scenarios when the endpoint returns `success === false` so the only type of error that could happen here is one that would cause `apiFetch` to throw, in which case it would be handled by the caller. I added a short sanity check to throw if the returned nonce is falsy. Later on we should distinguish between different types of errors and maybe even point specific menu items that triggered them.

noisysocks commented on packages/edit-navigation/src/store/actions.js:
```diff
@@ -0,0 +1,274 @@
+			post.blocks[ 0 ]
+		);
+		if ( ! response.success ) {
+			throw new Error();
```
Is throwing necessary? Looks like we can use `if ( response.success )` to determine what action we dispatch.

adamziel replied:
There are two different types of error that may happen:

1. Response has code 200 and body indicating `success === false`. In this case `apiFetch` just returns the response for us to handle.
2. Something unpredicted happens, like HTTP response code 500. In this case `apiFetch` throws an error.

For now we handle errors by just saying "there was an error" without communicating the root cause. Throwing allows us to handle both types of errors in the `catch` clause. I think it's fine, at least until we implement more sophisticated/granular error handling.

noisysocks replied:
> I think it's fine, at least until we implement more sophisticated/granular error handling.

Can you open an issue for this?

noisysocks commented on packages/edit-navigation/src/store/reducer.js:
```diff
@@ -0,0 +1,66 @@
+		case 'POP_PENDING_ACTION':
+			const postState = state[ postId ] || {};
+			if ( 'pendingActions' in postState ) {
+				postState.pendingActions = postState.pendingActions?.filter(
```
A reducer shouldn't mutate the `state` object directly which is what this will do because `postState` may be a reference to `state[ postId ]`.

adamziel replied:
Good spot! I missed that one.

noisysocks commented on packages/edit-navigation/src/store/selectors.js:
```diff
@@ -0,0 +1,57 @@
+import { KIND, POST_TYPE, buildNavigationPostId } from './utils';
+
+/**
+ * @see resolvers.js
```
I think it's worth repeating the doc comment for clarity and because the `@return` is different.

noisysocks reviewed (commented):
This is looking really good! 🌈 ✨ 

I tested the Navigation screen locally and everything seems to be working as before. It's really cool to see that <kbd>Cmd</kbd>+<kbd>Z</kbd> will undo a change. I wasn't able to get <kbd>Cmd</kbd>+<kbd>Shift</kbd>+<kbd>Z</kbd> to work, though.

Before merging, let's:

- Improve the error handling situation in `saveNavigationPost()`.
- Add JSDoc comments to every exported identifier in `store/**/*.js`. (I know, I know.)

I'd also like unit tests for every exported function in `store/**/*.js`, but that can be done in a follow-up PR.

adamziel commented on packages/edit-navigation/src/components/menus-editor/index.js:
```diff
@@ -16,7 +16,7 @@ import { __ } from '@wordpress/i18n';
  */
 import CreateMenuArea from './create-menu-area';
-import MenuEditor from '../menu-editor';
+import NavigationEditor from '../navigation-editor';
```
potentially, but not in this PR

adamziel committed:
Update comment on getNavigationPost selector

adamziel committed:
Update comment on getNavigationPost

adamziel committed:
Ensure POP_PENDING_ACTION action don't mutate the state

adamziel committed:
Rename getNavigationPost to getNavigationPostForMenu

adamziel committed:
Handle empty nonce in batchSave

adamziel committed:
Don't export storeConfig

adamziel commented on packages/edit-navigation/src/index.js:
```diff
@@ -63,3 +63,5 @@ export function initialize( id, settings ) {
 	);
 }
+
+export { storeConfig } from './store';
```
It doesn't - it's just a dev artifact :-)

adamziel committed:
Simplify useCallback

adamziel committed:
Rename NavigationEditorWrapper to NavigationEditor

adamziel commented on packages/edit-navigation/src/components/navigation-editor/index.js:
```diff
@@ -0,0 +1,78 @@
+	onDeleteMenu,
+} ) {
+	const { post, hasResolved } = useSelect( ( select ) => ( {
+		post: select( 'core/edit-navigation' ).getNavigationPost( menuId + '' ),
```
The parent component (`MenusEditor`) is not consistent with the type of menuId prop it passes here, after some more thought I fixed it there and made it always a number.

adamziel committed:
Ensure menuId is a number

adamziel committed:
Add comments for all exported members of store/**/*.js

adamziel committed:
Add comments to exported members of utils.js

adamziel commented:
@noisysocks I addressed all your feedback, feel free to merge if you think this one looks good now.

noisysocks reviewed (approved):
Nice work here! 🌈👍

You probably have to run `npm run docs:build` to make Travis CI happy.

adamziel committed:
Update package-lock.json
